docs/WAIVERS.md

```diff
@@ -137,17 +108,87 @@ returning `False` if the RPM is not installed.
     rhel == 8 and 'some thing' in note and env('INFRA') == 'jenkins'
 ```
 
-### Expecting both `pass` and `fail`/`error`
+## Collecting a list of sections
 
-Sometimes, failures or errors are not reliable and happen only occassionally.
-This is why waivers are "permissive" by default - they only waive
-`fail`/`error`, but allow `pass` to pass through, only noting `waived pass`
-in the result note.
+Sections (as defined above) are gathered from multiple waiver files
+in alphanumerical order, and - within each file - from the top.
 
-To override this, pass `strict=True` to `Match()`, which tells the waiving code
-to trigger a waive failure when the waiver matches `pass`.
+This creates a one big unified sequential ordered list of sections
+that is later used for waiving.
 
-```python
-/some/result/name
-    Match(rhel >= 8, strict=True)
+## How a result is waived
+
+When `results.report()` is called by test code to report a new result,
+the waiving logic looks at the status (`pass`, `fail`, etc.).
+
+If the status is `info` or `warn`, no further processing is done, and
+the result remains intact.  
+All other statuses continue below.
+
+The waiving logic then goes through the big list of sections (gathered
+above), **from beginning to end**, and tries to match the test name
+contained within the result, against all regexps in each section.
+
+For example, a test name of `/some/result/name`, when evaluated against
+```
+/different/name.*
+    True
+
+/some/unrelated/string
+/some/.+/name
+    True
 ```
+would match the second section, and its `/some/.+/name` regexp.
+
+**If no section matches on a regexp, no further processing is done,
+and the result remains intact.**
+
+When a section is matched on regexp, its python expression is evaluated.
+If that expression returns `False`, the section is skipped, and further
+sections (in the big unified list of sections) are tried, to see if
+at least one of them could match.
+
+**If no section matches on both a regexp AND python expression returning
+True, no further processing is done, the result remains intact.**
+
+When the result matches on both regexp + python of one section, the result
+is subject to waiving, and **no further sections are evaluated for that
+result**, the processing stops on the first matching section.
+
+If a result got through all that, the waiving logic decides based on the
+reported status (`pass`, `fail`, etc.):
+
+- `fail` is changed to `warn`, and "waived fail" is added to `note`
+- `error` is changed to `warn`, and "waived error" is added to `note`
+
+Further, if a waiver returns `strict=True` via `Match()`, or if
+`CONTEST_STRICT_WAIVERS` is set:
```

Comment:
Add extra info about permissive(default) / strict

```suggestion
By default, waivers are permissive causing `pass` result is
reported as `pass (waived pass)` result.

That might be changed to strict via waiver `strict=True`, e.g.
`Match(rhel >= 8, strict=True)`, or running Contest with
`CONTEST_STRICT_WAIVERS`. Adds waiving logic for `pass`:
```

Reply:
Again, I think I like my version better - it doesn't try to make `pass` a special free-form text thing. I'm not sure, "Adds waiving logic for `pass`"  is clearer here.

And `pass (waived pass)` also doesn't appear anywhere.

But I agree that `Match(rhel >= 8, strict=True)` should probably be mentioned, since I removed the section explaining how to use `strict=True`.

Reply:
Sure, you don't have to use exactly what I wrote. But I'd like to have some parts at least clearer.
For example:
```
By default, waivers are permissive causing if `pass` result matches a waiver, the result is intact.
```
or
```
By default, waivers are permissive causing if `pass` result matches a waiver, no `fail`/`error` is reported.
```
"Adds waiving logic for `pass`" can be also changed to something better like "Changes `pass` waiving behavior".